components/script/dom/servoparser/async_html.rs

```diff
@@ -47,171 +179,130 @@ impl Tokenizer {
-            },
+        let mut send_tendrils = Vec::new();
+        while let Some(str) = input.pop_front() {
+            send_tendrils.push(SendTendril::from(str));
+        }
+        // The input contains VecDeque<StrTendril>, which is converted to
+        // Vec<SendTendril>, so that it can be shared between threads with the use of a mutex.
+        // I used a mutex here because we need to access `send_tendrils` again when we receive the
+        // result of the feed function from the HtmlTokenizer. We use it to update the `input` parameter.
+        let send_tendrils = Arc::new(Mutex::new(send_tendrils));
+
+        // Send message to parser thread, asking it to start reading from the shared input.
+        // Parser operation messages will be sent to main thread as they are evaluated.
+        self.html_tokenizer_sender.send(ToHtmlTokenizerMsg::Feed(send_tendrils.clone())).unwrap();
+
+        loop {
+            match self.receiver.recv().expect("Unexpected channel panic in main thread.") {
+                ToTokenizerMsg::ProcessOperation(parse_op) => self.process_operation(parse_op),
+                ToTokenizerMsg::IsSameTree(ref x_id, ref y_id) => {
+                    let x = self.get_node(x_id);
+                    let y = self.get_node(y_id);
+
+                    let x = x.downcast::<Element>().expect("Element node expected");
+                    let y = y.downcast::<Element>().expect("Element node expected");
+                    self.sink_sender.send(ToSinkMsg::IsSameTree(x.is_in_same_home_subtree(y))).unwrap();
+                },
+                ToTokenizerMsg::HasParentNode(ref id) => {
+                    let res = self.get_node(id).GetParentNode().is_some();
+                    self.sink_sender.send(ToSinkMsg::HasParentNode(res)).unwrap();
+                },
+                ToTokenizerMsg::TokenizerResultDone => return Ok(()),
+                ToTokenizerMsg::TokenizerResultScript(script) => {
+                    let script = self.get_node(&script.id);
+                    let buffers = (*send_tendrils.lock().unwrap()).iter().
+                        map(|st| StrTendril::from(st.clone())).collect::<VecDeque<StrTendril>>();
+                    let buffer_queue = create_buffer_queue(buffers);
+
+                    // Update `input` parameter
+                    *input = buffer_queue;
+                    return Err(Root::from_ref(script.downcast().unwrap()));
+                }
+                ToTokenizerMsg::End => unreachable!(),
+            };
         }
     }
 
     pub fn end(&mut self) {
-        self.inner.end();
+        self.html_tokenizer_sender.send(ToHtmlTokenizerMsg::End).unwrap();
+        loop {
+            match self.receiver.recv().expect("Unexpected channel panic in main thread.") {
+                ToTokenizerMsg::ProcessOperation(parse_op) => self.process_operation(parse_op),
+                ToTokenizerMsg::IsSameTree(ref x_id, ref y_id) => {
```

Comment:
@nox The code for this match arm (and for the one below) is exactly the same as https://github.com/servo/servo/pull/17565/files#diff-10b46cb1e26142d2058e291de25bd4c7R249 . Should I extract them into methods and call them, or should I leave them be?